### PR TITLE
[Ready for Review] Update User Settings UI

### PR DESCRIPTION
**Pull request by samchrisinger**
## Purpose

**Proposal: https://docs.google.com/document/d/1oEED-vGOyC0K5CjdVpBqgtdGnO0021ICNitzim1AhKs/edit**

Making users check a checkbox to enable addons from the user settings page is slow and confusing. This PR removes those checkboxes, and instead lists each addon show the 0 or more accounts connected as well as the 0 or more nodes connected to that account.

Also, grouping twofactor with other addons doesn't make much sense, this moves twofactor to the account settings pane.

## Changes

Before we checked if the {Addon}UserSettings model was instantiated before rendering the addon's user config. This PR lets those models be uninstatiated. Connecting an account is equivalent to enabling the addon (for a user). 

Also, the twofactor settings UI has been updated to be asnyc.  Backend changes to support this are included as well.

## Side Effects

We'll need to regression test user/node settings workflows. 

Also, this includes significant UI changes that will likely require updates to our user documentation.

### CODE REVIEW

**Comment by samchrisinger** [comment]
@GaryKriebel @lbanner,
Please be aware that this PR includes significant UI changes, in particular with user settings. This will require updating user documentation.

**Comment by samchrisinger** [comment]
@GaryKriebel @lbanner , do we want to target this for 0.32?

**Comment by sloria** [comment]
Putting this in the queue for 0.33.0

**Comment by lbanner** [comment]
@sloria and @icereval : Have you two reviewed and added any needed comments to @samchrisinger 's proposal? Please do so if not -- Linked here: https://docs.google.com/a/cos.io/document/d/1oEED-vGOyC0K5CjdVpBqgtdGnO0021ICNitzim1AhKs/edit?usp=sharing  ... thanks.

**Comment by samchrisinger** [comment]
@sloria, @lbanner, @GaryKriebel 
Tis

**Comment by lbanner** [comment]
Excellent! Thanks -

File: website/addons/box/__init__.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 ADDED_MANDATORY = []
 
 VIEWS = []
-CONFIGS = ['user', 'node']
+CONFIGS = ['accounts', 'node']
```
Comment (sloria) [inline website/addons/box/__init__.py]: Why was this changed? What is the difference between `users` and `accounts` configs?

**Comment by lbanner** [comment]
@sloria ... want @lyndsysimon first pass here?

File: website/addons/box/__init__.py (hunk 1 above)
Comment (samchrisinger) [inline website/addons/box/__init__.py]: ```users``` config is a config related to one and only one user (i.e. twofactor), whereas ```accounts``` pertains only to external-account style addons that may be shared by multiple users.

File: website/addons/figshare/templates/figshare_user_settings.mako [hunk 2]
```diff
@@ -1,14 +1,15 @@
 <!-- Authorization -->
 <div>
     <h4 class="addon-title">
+      <img class="addon-icon" src="${addon_icon_url}"></img>
         figshare
         <small class="authorized-by">
             % if authorized:
-                    authorized
-                <a id="figshareDelKey" class="text-danger pull-right addon-auth">Delete Access Token</a>
+                    authorized by <a href="{{ profileUrl }}"><em> ${name} </em></a>
```
Comment (sloria) [inline website/addons/figshare/templates/figshare_user_settings.mako]: To prevent bots from crawling `{{ profileUrl }}`, use `data-bind="attr: {href: profileUrl}"`

Comment (sloria) [inline website/addons/figshare/templates/figshare_user_settings.mako]: This link doesn't seem to be working. I get the following 404 when clicking the link.

**Comment by sloria** [comment]
I'm not sure I agree with this language change:

osf___configure_external_accounts

Was this requested by Community?

File: website/addons/figshare/templates/figshare_user_settings.mako (hunk 2 above)
Comment (sloria) [inline website/addons/figshare/templates/figshare_user_settings.mako]: It appears to incorrectly link to `http://localhost:5000/settings/addons/%7B%7B%20profileUrl%20%7D%7D`. I think using the explicit `data-bind` should fix this.

**Comment by sloria** [comment]
jquery-qrcode need to be added to package.json.

File: website/addons/s3/templates/s3_user_settings.mako [hunk 3]
```diff
@@ -2,12 +2,13 @@
...
     <div>
         <h4 class="addon-title">
+          <img class="addon-icon" src="${addon_icon_url}"></img>
             Amazon S3
 
             <small class="authorized-by">
                 % if has_auth:
-                    authorized
-                    <a id="s3RemoveAccess" class="text-danger pull-right addon-auth">Delete Credentials</a>
+                    authorized by <a href="{{ profileUrl }}"><em>${name}</em></a></th>
```
Comment (sloria) [inline website/addons/s3/templates/s3_user_settings.mako]: Same here; use `data-bind="attr:...."`

File: website/addons/twofactor/static/test/twoFactorUserConfig.test.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        otpauth: faker.internet.ip(),
+        disable: faker.internet.ip()
+    }
+};
+var ENABLED_AND_CONFIRMED = {
+
+};
+var SELECTOR = '#myQRCode';
+
+describe('Two-factor User Config', () => {
```
Comment (sloria) [inline website/addons/twofactor/static/test/twoFactorUserConfig.test.js]: Thank you for testing.

File: website/addons/twofactor/static/twoFactorUserConfig.js [hunk 5]
```diff
@@ -1,63 +1,204 @@
...
+ViewModel.prototype.updateFromData = function(data) {
+    var self = this;
+    self.isEnabled(data.is_enabled);
+    self.isConfirmed(data.is_confirmed);
+    self.secret(data.secret);
+    self.urls = data.urls;
+    if (self.isEnabled()) {
+        // Initialize QR Code
+        if (!self.urls.otpauth) {
+            throw new Error('Tried to initailize jQuery.fn.qrcode without a otpauth URL');
```
Comment (sloria) [inline website/addons/twofactor/static/twoFactorUserConfig.js]: "initialize"

File: website/addons/twofactor/static/twoFactorUserConfig.js [hunk 6]
```diff
@@ -1,63 +1,204 @@
...
+            });
+            self.changeMessage('Could not retrieve Two-factor settings at ' +
+                'this time. Please refresh the page. ' +
+                'If the problem persists, email ' +
+                '<a href="mailto:support@osf.io">support@osf.io</a>.', 'text-danger', 5000);
+        });
+};
+
+/** Change the flashed message. */
+ViewModel.prototype.changeMessage = function(text, css, timeout) {
```
Comment (sloria) [inline website/addons/twofactor/static/twoFactorUserConfig.js]: Might be worth inheriting from `ChangeMessageMixin` to remove this duplicated code.

File: website/addons/twofactor/views.py [hunk 7]
```diff
@@ -25,3 +26,37 @@ def user_settings(user_addon, *args, **kwargs):
...
+        'result': serialize_settings(auth),
+    }
+
+
+@must_be_logged_in
+def twofactor_enable(auth, *args, **kwargs):
+    user = auth.user
+
+    if user.has_addon('twofactor'):
+        return HTTPError(http.BAD_REQUEST, message='This user already has twofactor enabled')
```
Comment (sloria) [inline website/addons/twofactor/views.py]: The correct way to pass custom error messages is by passing the `data` argument:

```python
HTTPError(http.BAD_REQUEST, data=dict(message_long='This user already has twofactor enabled'))
```

Comment (sloria) [inline website/addons/twofactor/views.py]: Also, keep the casing and hyphenation consistent in messages passed to the client; I believe I saw "Two-factor" in most other places, but make sure they're consistent.

File: website/addons/twofactor/views.py [hunk 8]
```diff
@@ -25,3 +26,37 @@ def user_settings(user_addon, *args, **kwargs):
...
+@must_be_logged_in
+@must_have_addon('twofactor', 'user')
+def twofactor_disable(auth, *args, **kwargs):
+
+    if auth.user.delete_addon('twofactor', auth=auth):
+        auth.user.save()
+        return {}
+    else:
+        raise HTTPError(http.INTERNAL_SERVER_ERROR, data=dict(
+            message='Could not disable twofactor at this time'
```
Comment (sloria) [inline website/addons/twofactor/views.py]: The key should be `message_long` here. Also check casing and hyphenation of "twofactor"

File: website/profile/views.py [hunk 9]
```diff
@@ -383,20 +363,28 @@ def user_notifications(auth, **kwargs):
...
-    for addon in addons:
-        js_path = paths.resolve_addon_path(addon.config, 'user-cfg.js')
+    for addon_config in addon_configs:
+        js_path = paths.resolve_addon_path(addon_config, 'user-cfg.js')
         if js_path:
             js_modules.append(js_path)
     return js_modules
 
 @must_be_logged_in
+def profile_addons(**kwargs):
```
Comment (sloria) [inline website/profile/views.py]: I don't see a route for this view. Remove this?

File: website/static/js/addonPermissions.js [hunk 10]
```diff
@@ -17,7 +17,7 @@ var AddonPermissionsTable = {
             var apiUrl = $(this).attr('api-url')+ addonShortName + '/config/';
             bootbox.confirm({
                 title: 'Remove addon?',
-                message: 'Are you sure you want to remove the ' + addonFullname + ' authorization from this project?',
+                message: 'Are you sure you want to disconnnect the ' + addonFullname + ' Account from this project?',
```
Comment (sloria) [inline website/static/js/addonPermissions.js]: "Account" should not be capitalized.

File: website/static/js/addonPermissions.js [hunk 11]
```diff
@@ -38,7 +38,7 @@ var AddonPermissionsTable = {
                             },
 
                             error: function () {
-                                $osf.growl('An error occurred, the project has not been deauthorized. ',
+                                $osf.growl('An error occurred, the Account is still connected to the project. ',
```
Comment (sloria) [inline website/static/js/addonPermissions.js]: Same here; don't capitalize "account". Also, it might be nice to add the provider name here.

File: website/static/js/folderPickerNodeConfig.js [hunk 12]
```diff
@@ -81,13 +81,13 @@ var FolderPickerViewModel = oop.defclass({
         self.messages = {
             invalidCredOwner: ko.pureComputed(function() {
                 return 'Could not retrieve ' + self.addonName + ' settings at ' +
-                    'this time. The ' + self.addonName + ' addon credentials may no longer be valid.' +
-                    ' Try deauthorizing and reauthorizing ' + self.addonName + ' on your <a href="' +
-                    self.urls().settings + '">account settings page</a>.';
+                    'this time. The credentials associated with this ' + self.addonName + ' Account may no longer be valid.' +
```
Comment (sloria) [inline website/static/js/folderPickerNodeConfig.js]: "account"

File: website/static/js/folderPickerNodeConfig.js [hunk 13]
```diff
@@ -81,13 +81,13 @@ var FolderPickerViewModel = oop.defclass({
...
+                    ' Try disconnecting and reconnecting the ' + self.addonName + 'Account on your <a href="' +
```
Comment (sloria) [inline website/static/js/folderPickerNodeConfig.js]: "account"

**Comment by sloria** [comment]
Doesn't necessarily need to be done before this is merged, but the logo images are rather large; it would be good to optimize those to improve page load speeds.

File: website/addons/dropbox/views/auth.py [hunk 14]
```diff
@@ -151,25 +150,32 @@ def dropbox_user_config_get(user_addon, auth, **kwargs):
...
+
+        if user_addon.has_auth:
+            try:
+                client = get_client_from_user_settings(user_addon)
+                client.account_info()
+            except ErrorResponse as error:
+                if error.status == 401:
+                    valid_credentials = False
+                else:
+                    raise HTTPError(http.BAD_REQUEST)
```
Comment (sloria) [inline website/addons/dropbox/views/auth.py]: Might be good to add a `message_long` here; can we get any info from the `ErrorResponse`?

File: website/templates/profile/addons.mako [hunk 15]
```diff
@@ -1,8 +1,14 @@
...
 <%def name="content()">
+<style>
+.addon-icon {
+    width: 20px;
+}
+</style>
+
 <% from website import settings %>
-<h2 class="page-header">Configure Add-ons</h2>
+<h2 class="page-header">Configure External Accounts</h2>
```
Comment (sloria) [inline website/templates/profile/addons.mako]: I think I mentioned this in a previous comment, but I think Configure External Accounts could be changed to "Configure Add-on Accounts" or "Configre Add-on External Accounts" for clarity

**Comment by sloria** [comment]
Pass finished. :full_moon_with_face:

**Comment by samchrisinger** [comment]
@GaryKriebel @lbanner I'm not going to spend any time on this until asked to do so.

**Comment by lbanner** [comment]
I am ready to see this move forward to QA --> Production. If it is competing with Pre-Reg., is there another Dev that might carry this forward?

**Comment by samchrisinger** [comment]
@lbanner  changes are mostly stylistic in nature-- most any dev here could do this. Estimate at most 2hrs work. 

But yes, this competes directly with Prereg development.

**Comment by GaryKriebel** [comment]
I just spoke with @samanehsan and she will pick up the updates of code review and then be available for bug fixes (if any) after it goes to QA.

**Comment by lbanner** [comment]
Great :+1: ... Thank you @samanehsan

**Comment by samanehsan** [comment]
@samchrisinger, @sloria is this the jquery-qrcode library being used? https://github.com/jeromeetienne/jquery-qrcode

**Comment by samchrisinger** [comment]
Looks to be, yes.

**Comment by samanehsan** [comment]
Done responding to code review :palm_tree: New PR is here:  https://github.com/CenterForOpenScience/osf.io/pull/3409 @sloria @samchrisinger

**Comment by sloria** [comment]
Closing in favor of https://github.com/CenterForOpenScience/osf.io/pull/3409
